crates/matrix-sdk/src/room/calls.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright 2025 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//!  Facilities to handle incoming calls.
+
+use ruma::{
+    events::{
+        rtc::decline::{RtcDeclineEventContent, SyncRtcDeclineEvent},
+        AnySyncMessageLikeEvent, AnySyncTimelineEvent,
+    },
+    EventId, OwnedUserId, UserId,
+};
+use thiserror::Error;
+use tokio::sync::broadcast;
+use tracing::instrument;
+
+use crate::{event_handler::EventHandlerDropGuard, room::EventSource, Room};
+
+/// An error occurring while interacting with a call/rtc event.
+#[derive(Debug, Error)]
+pub enum CallError {
+    /// We couldn't fetch the remote notification event.
+    #[error("Couldn't fetch the remote event: {0}")]
+    Fetch(Box<crate::Error>),
+
+    /// We tried to decline an event which is not of type m.rtc.notification.
+    #[error("You cannot decline this event type.")]
+    BadEventType,
+
+    /// We tried to decline a call started by ourselves.
+    #[error("You cannot decline your own call.")]
+    DeclineOwnCall,
+
+    /// We couldn't properly deserialize the target event.
+    #[error(transparent)]
+    Deserialize(#[from] serde_json::Error),
+}
+
+impl Room {
+    /// Create a new decline call event for the target notification event id .
+    ///
+    /// The event can then be sent with [`Room::send`] or a
+    /// [`crate::send_queue::RoomSendQueue`].
+    #[instrument(skip(self), fields(room = %self.room_id()))]
+    pub async fn make_decline_call_event(
+        &self,
+        notification_event_id: &EventId,
+    ) -> Result<RtcDeclineEventContent, CallError> {
+        make_call_decline_event(self, self.own_user_id(), notification_event_id).await
+    }
+
+    /// Subscribe to decline call evvent for this room.
+    ///
+    /// The returned receiver will receive the sender UserID for each decline
+    /// for the matching notify event.
+    pub fn subscribe_to_call_decline_events(
+        &self,
+        notification_event_id: &EventId,
+    ) -> (EventHandlerDropGuard, broadcast::Receiver<OwnedUserId>) {
```

Comment:
Can we get an example on in the docs how this is supposed to be used?

Reply:
added